### PR TITLE
[libc++] Make _LIBCPP_ASSUME usable when it is appropriate

### PR DESCRIPTION
**Pull request by davidben**
libc++ turned off _LIBCPP_ASSUME because turning every debug assert into __builtin_assume tripped
https://discourse.llvm.org/t/llvm-assume-blocks-optimization/71609

However, this means we can't use _LIBCPP_ASSUME when there is a clear optimization intent. See https://github.com/llvm/llvm-project/pull/78929#issuecomment-1936582711 for discussion of a place where _LIBCPP_ASSUME would be valuable.

Go ahead and fix this now, and adjust the comments. I don't think we want _LIBCPP_ASSUME in disabled asserts anyway, at least not in any of the hardened modes. This PR should have no impact on libc++ right now, since _LIBCPP_ASSUME is currently never called standalone, but I figure I can do this as a standalone PR since it's pretty straightforward.

### CODE REVIEW

**Review by ldionne — APPROVED** [review]
This makes sense to me. Deferring to @var-const for final approval, but at first glance I like this.

**Comment by davidben** [comment]
Friendly ping on this PR. This is a blocker to fixing #101370 which, in turn, is a blocker to making bounded iterators practical.

File: libcxx/include/__assert [hunk 1]
```diff
@@ -23,17 +23,20 @@
...
+// discussion.
+#if __has_builtin(__builtin_assume)
 #  define _LIBCPP_ASSUME(expression)                                                                                   \
     (_LIBCPP_DIAGNOSTIC_PUSH _LIBCPP_CLANG_DIAGNOSTIC_IGNORED("-Wassume")                                              \
          __builtin_assume(static_cast<bool>(expression)) _LIBCPP_DIAGNOSTIC_POP)
 #else
 #  define _LIBCPP_ASSUME(expression) ((void)0)
 #endif
 
+// Historically, disabled assertions below expanded to `_LIBCPP_ASSUME`, but this both triggers the
```
Comment (var-const) [inline libcxx/include/__assert:37]: Nit: I'd remove this comment -- I understand the desire to make sure somebody doesn't reintroduce the previous state (which seems to make a lot of sense in theory but doesn't work in practice), but IMO the comment we have on `_LIBCPP_ASSUME` should be enough of a deterrence.

**Review by var-const — APPROVED** [review]
So sorry about the very slow review! LGTM.

File: libcxx/include/__assert (hunk 1 above)
Comment (davidben) [reply libcxx/include/__assert]: Removed and rebased.

**Pull request merged**